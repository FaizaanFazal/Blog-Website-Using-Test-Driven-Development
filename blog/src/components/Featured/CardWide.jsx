import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';
import images from '../../utils/images';
import '../../styles/Card.scss';

const CardWide = ({ blogItemData }) => {
  const [author, setAuthor] = useState();
  const { authorId } = blogItemData;
  const [slicedDate, setSlicedDate] = useState('');

  const MAX_LENGTH = 250;
  const users = useSelector((state) => state.user.allusers);

  let filtered;
  useEffect(() => {
    console.log(users);
    filtered = users?.filter((user) => user.id === authorId);
    setAuthor(filtered[0]);
    if (filtered) {
      const dateString = blogItemData?.createdAt;
      setSlicedDate(dateString.substring(0, 10));
    } else {
      setSlicedDate('NA');
    }
  }, [users]);
  return (
    <div className="card-grid grid">
      <div className="card-grid-img">
        <img data-testid="FeaturedImg" src={`${blogItemData?.imageSrc}`} alt={`${blogItemData?.imageAlt}`} className="object-fit-cover" />
      </div>
      <div className="card-grid-text">
        <div>
          <div className="badge badge-primary" >featured</div>
          <h2 data-testid="headingfeatured" className="title-lg title featured-title">
            { blogItemData?.title }
          </h2>

          {blogItemData?.content?.length > MAX_LENGTH
            ? (
              <p className="text-lg text" data-testid="featureSummary">
                {`${blogItemData.content.substring(0, MAX_LENGTH)}...`}
                <Link data-testid="slugLink" to={`blogdetails/${blogItemData?.slug}`}>Read more</Link>
              </p>
            )
            : <p data-testid="featureSummary">{blogItemData?.content}</p>}
        </div>

        <div className="card-footer flex justify-between items-center">
          <div className="writer-info grid">
            <div className="info-avatar">
              <img src={`${author?.authorImageSrc}`} alt={`${author?.authorImageAlt}`} className="object-fit-cover" />
            </div>
            <div className="info-intro">
              <p className="intro-name text-base">{ author?.userName }</p>
              {author?.verified
                ? (
                  <div className="intro-verify flex items-center">
                    <img src={`${images.verify_icon_filled}`} alt="icon" className="verify-icon" />
                    <span className="verify-status text-sm text">Verified writer</span>
                  </div>
                )
                : (
                  <div className="intro-verify flex items-center">
                    <img src={`${images.verify_icon_light}`} alt="icon" className="verify-icon" />
                    <span className="verify-status text-sm text">Not-Verified writer</span>
                  </div>
                )}
            </div>
          </div>
          <div className="date-info text text-base" data-testid="featureDate">{ slicedDate}</div>
        </div>
      </div>
    </div>
  );
};

export default CardWide;
