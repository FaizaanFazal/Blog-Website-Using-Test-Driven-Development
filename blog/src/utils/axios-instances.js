// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
