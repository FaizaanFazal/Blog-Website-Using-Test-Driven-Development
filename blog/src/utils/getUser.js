export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };