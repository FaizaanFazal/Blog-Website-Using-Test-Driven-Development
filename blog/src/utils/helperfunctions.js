import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperName = (username) => {
  const regex = /^[A-Za-z]+$/;
  return regex.test(username);
};
export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[A-Z])(?=.*[0-9])(?=.*[^a-zA-Z0-9]).{8,}|^(?=.*[0-9])(?=.*[A-Z])(?=.*[^a-zA-Z0-9]).{8,}|^(?=.*[^a-zA-Z0-9])(?=.*[A-Z])(?=.*[0-9]).{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    authorId: 'de2f4aad-949e-4452-8799-aa09630b8e02',
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    imageSrc: images.blogImg1,
    imageAlt: '10 Tips for Budget-Friendly Travel Blog Header Image',
    verified: true,
    createdAt: '2023-11-10T11:40:54.577Z',
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    authorId: 'de2f4aad-949e-4452-8799-aa09630b8e02',
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    imageSrc: images.blogImg1,
    imageAlt: 'Elinore',
    verified: true,
    slug: 'post1',
    createdAt: '2023-11-10T11:40:54.577Z',
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
