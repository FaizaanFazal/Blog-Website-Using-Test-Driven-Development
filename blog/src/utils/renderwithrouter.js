import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}