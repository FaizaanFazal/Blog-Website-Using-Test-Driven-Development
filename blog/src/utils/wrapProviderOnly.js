import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

const AllTheProviders = ({ children }) => (
  <Provider store={reduxStore}>
    {children}
  </Provider>
);

export const renderWithProvidersOnly = (ui, options) => render(<AllTheProviders>{ui}</AllTheProviders>, options);

// export function renderWithProviders(
//   ui,
//   {
//     // preloadedState = {},
//     // Automatically create a store instance if no store was passed in
//     store = reduxStore,
//     ...renderOptions
//   } = {},
// ) {
//   function Wrapper({ children }) {
//     return <Provider store={store}>{children}</Provider>;
//   }

//   // Return an object with the store and all of RTL's query functions
//   return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
// }

renderWithProvidersOnly.propTypes = {
  children: PropTypes.node.isRequired,
};
