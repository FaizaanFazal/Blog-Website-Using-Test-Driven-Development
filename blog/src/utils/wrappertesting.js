import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
