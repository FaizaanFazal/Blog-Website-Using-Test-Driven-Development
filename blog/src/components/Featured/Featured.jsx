import React from 'react'
import CardWide from './CardWide'
import {useSelector} from 'react-redux'


export default function Featured() {

  const featured=useSelector(state=>state.blog.featured);

  return (
    <div role='featured' className="container">
        <section className="featured-sc bg-white">
        {featured.map(blogItem =>(<CardWide blogItemData = {blogItem}/>))}
       
        </section>
      </div>
  )
}
