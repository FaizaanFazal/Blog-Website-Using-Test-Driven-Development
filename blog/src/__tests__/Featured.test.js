import { screen, within } from '@testing-library/react';
import Featured from '../components/Featured/Featured';
import { renderWithProviders } from '../utils/wrappertesting';
import { isProperImageURL, small, greater } from '../utils/helperfunctions';
import CardWide from '../components/Featured/CardWide';

describe('Unit tests for Featured component', () => {
  it('Dom test for rendering Featured', () => {
    renderWithProviders(<Featured />);
    const featureElement = screen.getByTestId('featured');
    expect(featureElement).toBeInTheDocument();
  });

  it('snapshot test to check Featured component', () => {
    const { asFragment } = renderWithProviders(<Featured />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Image renders and src is correct', () => {
    renderWithProviders(<Featured />);
    const image = screen.getByTestId('FeaturedImg');
    expect(image).toBeInTheDocument();
    const imageSrc = image.src;
    expect(isProperImageURL(imageSrc)).toBe(true);
  });

  it('Image Alternative (alt) is correct', () => {
    renderWithProviders(<CardWide blogItemData={greater[0]} />);
    const Featuredimage = screen.getByTestId('FeaturedImg');
    expect(Featuredimage).toHaveAttribute('alt', greater[0].image.alt);
  });

  it('Heading is rendered', () => {
    renderWithProviders(<Featured />);
    const heading = screen.getByTestId('headingfeatured');
    expect(heading).toBeInTheDocument();
  });

  it('Summary  with greater than 250 chars', async () => {
    renderWithProviders(<CardWide blogItemData={greater[0]} />);
    const summaryG = screen.getByTestId('featureSummary');
    const anchorTagElementsG = within(summaryG).queryAllByTestId('slugLink');
    expect(anchorTagElementsG.length).toBe(1);
  });

  it('Summary with less than 250 chars', () => {
    renderWithProviders(<CardWide blogItemData={small[0]} />);
    const summarys = screen.getByTestId('featureSummary');
    const anchorTagElements = within(summarys).queryAllByTestId('slugLink');
    expect(anchorTagElements.length).toBe(0);
  });

  it('Date is rendered', () => {
    renderWithProviders(<Featured />);
    const featureDate = screen.getByTestId('featureDate');
    expect(featureDate).toBeInTheDocument();
  });

  it('Check view more slug link to be correct', () => {
    renderWithProviders(<CardWide blogItemData={greater[0]} />);
    const slugLink = screen.getByTestId('slugLink');
    expect(slugLink).toBeInTheDocument();
  });
});
