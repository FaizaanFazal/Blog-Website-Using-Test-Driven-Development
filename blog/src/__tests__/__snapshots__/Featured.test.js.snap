// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for Featured component snapshot test to check Featured component 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="featured"
  >
    <section
      class="featured-sc bg-white"
    />
  </div>
</DocumentFragment>
`;
